File: agent/src/ipc.cpp
#include "ipc.h"

#include <fcntl.h>
#include <poll.h>
#include <sys/socket.h>
#include <sys/stat.h>
#include <sys/un.h>
#include <unistd.h>

#include <cstring>
#include <stdexcept>

namespace ipc {

std::string defaultSocketPath() {
    const char* rt = getenv("XDG_RUNTIME_DIR");
    std::string base = rt && *rt ? rt : "/run/user/" + std::to_string(getuid());
    return base + "/logimx.sock";
}

static bool writeAll(int fd, const std::string& s) {
    size_t off = 0;
    while (off < s.size()) {
        ssize_t w = ::send(fd, s.data() + off, s.size() - off, MSG_NOSIGNAL);
        if (w <= 0) return false;
        off += static_cast<size_t>(w);
    }
    return true;
}

Server::Server(Handler h, std::string path) : handler_(std::move(h)), path_(std::move(path)) {
    ::unlink(path_.c_str());
    fd_ = ::socket(AF_UNIX, SOCK_STREAM | SOCK_CLOEXEC, 0);
    if (fd_ < 0) throw std::runtime_error("socket");
    sockaddr_un addr{};
    addr.sun_family = AF_UNIX;
    strncpy(addr.sun_path, path_.c_str(), sizeof(addr.sun_path) - 1);
    if (::bind(fd_, reinterpret_cast<sockaddr*>(&addr), sizeof(addr)) < 0) throw std::runtime_error("bind " + path_ + ": " + strerror(errno));
    chmod(path_.c_str(), 0600);
    ::listen(fd_, 64);
    thread_ = std::thread([this] { acceptLoop(); });
}

Server::~Server() {
    stop_ = true;
    ::shutdown(fd_, SHUT_RDWR);
    ::close(fd_);
    if (thread_.joinable()) thread_.join();
    {
        std::lock_guard<std::mutex> lk(clientsMutex_);
        for (int c : clients_) ::shutdown(c, SHUT_RDWR);
    }
    ::unlink(path_.c_str());
}

void Server::acceptLoop() {
    while (!stop_) {
        pollfd p{fd_, POLLIN, 0};
        if (::poll(&p, 1, 300) <= 0) continue;
        for (;;) {  // drain everything pending
            int c = ::accept4(fd_, nullptr, nullptr, SOCK_CLOEXEC | SOCK_NONBLOCK);
            if (c < 0) break;
            int fl = fcntl(c, F_GETFL);
            if (fl >= 0) fcntl(c, F_SETFL, fl & ~O_NONBLOCK);
            {
                std::lock_guard<std::mutex> lk(clientsMutex_);
                clients_.insert(c);
            }
            std::thread([this, c] { serve(c); }).detach();
        }
    }
}

void Server::serve(int fd) {
    std::string buf;
    char tmp[4096];
    while (!stop_) {
        ssize_t n = ::recv(fd, tmp, sizeof(tmp), 0);
        if (n <= 0) break;
        buf.append(tmp, static_cast<size_t>(n));
        if (buf.size() > 8 * 1024 * 1024 && buf.find('\n') == std::string::npos) break;  // oversized request, drop client
        size_t nl;
        while ((nl = buf.find('\n')) != std::string::npos) {
            std::string line = buf.substr(0, nl);
            buf.erase(0, nl + 1);
            json resp, reqId;
            try {
                json req = json::parse(line);
                if (req.is_object()) reqId = req.value("id", json());
                json params = req.is_object() ? req.value("params", json::object()) : json::object();
                if (!params.is_object()) params = json::object();
                resp = {{"id", reqId}, {"result", handler_(req.is_object() ? req.value("method", "") : "", params)}};
            } catch (const std::exception& e) {
                resp = {{"id", reqId}, {"error", e.what()}};
            }
            if (!writeAll(fd, resp.dump() + "\n")) break;
        }
    }
    {
        std::lock_guard<std::mutex> lk(clientsMutex_);
        clients_.erase(fd);
    }
    ::close(fd);
}

void Server::broadcast(const std::string& event, const json& data) {
    std::string msg = json{{"event", event}, {"data", data}}.dump() + "\n";
    std::lock_guard<std::mutex> lk(clientsMutex_);
    for (int c : clients_) writeAll(c, msg);
}

Client::Client(std::string path) {
    fd_ = ::socket(AF_UNIX, SOCK_STREAM | SOCK_CLOEXEC, 0);
    sockaddr_un addr{};
    addr.sun_family = AF_UNIX;
    strncpy(addr.sun_path, path.c_str(), sizeof(addr.sun_path) - 1);
    if (::connect(fd_, reinterpret_cast<sockaddr*>(&addr), sizeof(addr)) < 0) {
        ::close(fd_);
        fd_ = -1;
        throw std::runtime_error("agent not running (" + path + ")");
    }
}

Client::~Client() {
    if (fd_ >= 0) ::close(fd_);
}

json Client::call(const std::string& method, const json& params) {
    json req = {{"id", ++id_}, {"method", method}, {"params", params}};
    if (!writeAll(fd_, req.dump() + "\n")) throw std::runtime_error("send failed");
    char tmp[4096];
    for (;;) {
        size_t nl = buf_.find('\n');
        if (nl != std::string::npos) {
            std::string line = buf_.substr(0, nl);
            buf_.erase(0, nl + 1);
            json resp = json::parse(line);
            if (resp.contains("event")) continue;  // ignore broadcasts on a request socket
            if (resp.contains("error")) throw std::runtime_error(resp["error"].get<std::string>());
            return resp["result"];
        }
        ssize_t n = ::recv(fd_, tmp, sizeof(tmp), 0);
        if (n <= 0) throw std::runtime_error("agent closed the connection");
        buf_.append(tmp, static_cast<size_t>(n));
    }
}

}  // namespace ipc

File: logimx/ipc.py
"""Tiny JSON-lines RPC over a UNIX socket, shared by daemon, CLI and UI."""
from __future__ import annotations

import json
import os
import socket
import threading
from typing import Callable

SOCKET_PATH = os.path.join(os.environ.get("XDG_RUNTIME_DIR", f"/run/user/{os.getuid()}"), "logimx.sock")


class Server:
    def __init__(self, handler: Callable[[str, dict], object], path: str = SOCKET_PATH):
        self.handler = handler
        self.path = path
        self._stop = threading.Event()
        try:
            os.unlink(path)
        except FileNotFoundError:
            pass
        self.sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
        self.sock.bind(path)
        os.chmod(path, 0o600)
        self.sock.listen(8)
        self.sock.settimeout(0.5)
        self._thread = threading.Thread(target=self._accept, daemon=True, name="ipc-accept")
        self._thread.start()

    def close(self):
        self._stop.set()
        try:
            self.sock.close()
            os.unlink(self.path)
        except OSError:
            pass

    def _accept(self):
        while not self._stop.is_set():
            try:
                conn, _ = self.sock.accept()
            except socket.timeout:
                continue
            except OSError:
                return
            threading.Thread(target=self._serve, args=(conn,), daemon=True).start()

    def _serve(self, conn: socket.socket):
        f = conn.makefile("rwb", buffering=0)
        try:
            for line in f:
                try:
                    req = json.loads(line)
                    result = self.handler(req.get("method", ""), req.get("params") or {})
                    resp = {"id": req.get("id"), "result": result}
                except Exception as e:  # report, keep serving
                    resp = {"id": None, "error": f"{type(e).__name__}: {e}"}
                f.write((json.dumps(resp, default=str) + "\n").encode())
        except (OSError, ValueError):
            pass
        finally:
            conn.close()


class Client:
    def __init__(self, path: str = SOCKET_PATH):
        self.path = path
        self.sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
        self.sock.settimeout(5.0)
        self.sock.connect(path)
        self.f = self.sock.makefile("rwb", buffering=0)
        self._id = 0

    def call(self, method: str, **params):
        self._id += 1
        self.f.write((json.dumps({"id": self._id, "method": method, "params": params}) + "\n").encode())
        while True:
            line = self.f.readline()
            if not line:
                raise ConnectionError("daemon closed the connection")
            resp = json.loads(line)
            if "event" not in resp:  # skip broadcasts on a request socket
                break
        if "error" in resp:
            raise RuntimeError(resp["error"])
        return resp["result"]

    def close(self):
        try:
            self.sock.close()
        except OSError:
            pass
